### DISCUSSION TITLE
Smooth Font Overwrite With Background Color

### DISCUSSION THREAD

**Discussion by barn53** [Bodmer/TFT_eSPI#1001] [Q&A - General] [answered]
Hi,

for a project I want to use the smooth font feature of TFT_eSPI which is great, by the way!
I need to regularly overwrite a numerical string by another.
The problem here is, the previous characters are not overwritten by the background color (black) before the new characters are written. What is the best way to solve this problem?

I put a black rectangle over the region, then write the string. But this looks too flickering.

I tried to write the string into a cleared to black sprite and pushed the sprite to the screen, which looks great. But then I have memory issues. So this is also no option.

I think a good approach would be that the library writes a rectangle in background color in the size of each character before it prints the character. Is there already such an option available?

I tried setTextPadding(), but saw no effect.

Any ideas are highly appreciated.

Thank you,
Gerd

**Comment by Bodmer** [comment]
See the "Font_Demo_x_Array" examples and read the header info as this explains things.

The best way to avoid flicker is to use a sprite of the correct size.

**Reply by barn53** [reply]
Thank you, but as described a sprite is no option, as the memory usage is too high.

**Comment by Bodmer** [comment]
Font_Demo_2_Array demonstrates the use of setTextPadding,(see line 99), you will get flicker if you are not able to use a sprite.

**Reply by barn53** [reply]
Hi, please have a look at my pull request https://github.com/Bodmer/TFT_eSPI/pull/1002 
It looks really great, there is no flicker to see.

**Reply by Bodmer** [reply]
Thanks, yes this will reduce flicker to 1 character at a time and hence will be less intrussive, see comments in the pull request.

**Reply by barn53** [reply]
The trick is, it reduces flicker to a horizontal line of 1 pixel height across one character while drawing the glyph. So the flickering area and time is way smaller than 1 character. But you are totally right, this approach is too naive as I did not think about overlapping fonts. So, when I find some time I will look into this. Regards.

**Comment by Bodmer** [comment]
I realised that my test should have reduced flicker more!

I looked at my changed code and spotted an error in the glyph blanking rectangle calculations. This has been corrected and flicker is negligible now. The right hand edge of the last blanking area is tracked so it should handle overlapping font characters.

By default this new blanking method will be off for the time being to avoid breaking the RTL code in the TFT_eFEX extension. To turn on flicker reduction use:
```
tft.setAttribute(SFBG_ENABLE, true);
```

I will update the version 2.4.0 beta branch later today and run some more tests.

**Comment by Bodmer** [comment, marked as answer]
The experimental changes have been pushed to the beta branch. This has had limited testing, the functionality will be retained but the implementation may be updated in the near future. Font_Demo_2_Array indicates good flicker reduction and correct blanking.
